### ISSUE TITLE
[1.1.1] Out-of-tree builds fail on `make update`

### ISSUE THREAD

**Issue by romen** [openssl/openssl#11940]
It seems like currently oot-builds succeed only if the build dir and the source dir are at the same level of nesting

### Steps to reproduce

```bash
; cd $osslrepo
; git checkout origin/OpenSSL_1_1_1-stable
; srcdir=$PWD
; blddir=../_build/nest/a/little/more
; mkdir -p $blddir
; cd $blddir
; $srcdir/config
; make update
```

Notice that if `$blddir` is set to `../_build`, the failure is avoided.

~~To quote @levitte: "I have a branch!" (where I can trigger the error in Travis): https://github.com/romen/openssl/tree/issues/1.1.1/oot_build~~

~~[Here](https://travis-ci.com/github/romen/openssl/builds/171098007) you can see the build passing using `../_build` as the build dir, [here](https://travis-ci.com/github/romen/openssl/builds/171097939) you can see the build failing using `../_build/nest/a/little/more` instead.~~
[Here](https://github.com/openssl/openssl/pull/14388/checks?check_run_id=2012180309) you can see the build failing in GitHub CI: #14388.

This seems to be related to #11184 and #11185, that addressed a similar defect in `master`.

_edit: updated to point to #14388 now that Travis CI is no more._

**Comment by romen** [comment]
@levitte do you have a suggestion?

**Comment by levitte** [comment]
Is this issue still current?

**Comment by romen** [comment]
I haven't tried recently, but my gut says yes! I am not near a computer at the moment to test!

**Comment by romen** [comment]
@levitte , you probably already verified this, but anyway:

```bash
romen@buildripper:/tmpram/111$ git checkout origin/OpenSSL_1_1_1-stable
HEAD is now at a88ea7dfdf Check ASN1_item_ndef_i2d() return value.
romen@buildripper:/tmpram/111$ srcdir=$PWD
romen@buildripper:/tmpram/111$ blddir=../_build/nest/a/little/more
romen@buildripper:/tmpram/111$ mkdir -p $blddir
romen@buildripper:/tmpram/111$ cd $blddir
romen@buildripper:/tmpram/_build/nest/a/little/more$ $srcdir/config
Operating system: x86_64-whatever-linux2
Configuring OpenSSL version 1.1.1k-dev (0x101010b0L) for linux-x86_64
Using os-specific seed configuration
Creating configdata.pm
Creating Makefile

**********************************************************************
***                                                                ***
***   OpenSSL has been successfully configured                     ***
***                                                                ***
***   If you encounter a problem while building, please open an    ***
***   issue on GitHub <https://github.com/openssl/openssl/issues>  ***
***   and include the output from the following command:           ***
***                                                                ***
***       perl configdata.pm --dump                                ***
***                                                                ***
***   (If you are new to OpenSSL, you might want to consult the    ***
***   'Troubleshooting' section in the INSTALL file first)         ***
***                                                                ***
**********************************************************************
romen@buildripper:/tmpram/_build/nest/a/little/more$ make update
( cd ../../../../../111; /usr/bin/perl VMS/VMSify-conf.pl \
                        < apps/openssl.cnf > apps/openssl-vms.cnf )
( cd ../../../../../111; /usr/bin/perl crypto/bn/bn_prime.pl > crypto/bn/bn_prime.h )
( cd ../../../../../111; /usr/bin/perl crypto/objects/objects.pl -n \
                        crypto/objects/objects.txt \
                        crypto/objects/obj_mac.num \
                        > crypto/objects/obj_mac.new && \
    mv crypto/objects/obj_mac.new crypto/objects/obj_mac.num )
( cd ../../../../../111; /usr/bin/perl crypto/objects/objects.pl \
                        crypto/objects/objects.txt \
                        crypto/objects/obj_mac.num \
                        > include/openssl/obj_mac.h )
( cd ../../../../../111; /usr/bin/perl crypto/objects/obj_dat.pl \
                        include/openssl/obj_mac.h \
                        > crypto/objects/obj_dat.h )
( cd ../../../../../111; /usr/bin/perl crypto/objects/objxref.pl \
                        crypto/objects/obj_mac.num \
                        crypto/objects/obj_xref.txt \
                        > crypto/objects/obj_xref.h )
( cd ../../../../../111; /usr/bin/perl crypto/conf/keysets.pl \
                        > crypto/conf/conf_def.h )
( cd ../../../../../111; /usr/bin/perl crypto/asn1/charmap.pl \
                        > crypto/asn1/charmap.h )
( cd ../../../../../111; /usr/bin/perl fuzz/mkfuzzoids.pl \
                        crypto/objects/obj_dat.h \
                        > fuzz/oids.txt )
( b=`pwd`; set -e; cd ../../../../../111; \
          /usr/bin/perl util/ck_errf.pl -strict -internal; \
          /usr/bin/perl -I$b util/mkerr.pl  -internal )
( b=`pwd`; set -e; cd ../../../../../111/engines; \
          for E in *.ec ; do \
              /usr/bin/perl ../util/ck_errf.pl -strict \
                -conf $E `basename $E .ec`.c; \
              /usr/bin/perl -I$b ../util/mkerr.pl  -static \
                -conf $E `basename $E .ec`.c ; \
          done )
( b=`pwd`; cd ../../../../../111; /usr/bin/perl -I$b util/mkdef.pl crypto update )
Can't open opensslv.h at util/mkdef.pl line 1497.
make: *** [Makefile:597: ordinals] Error 2
```

**Comment by romen** [comment]
Also, the issue appears also using absolute paths:

```bash
romen@buildripper:/tmpram/111$ git checkout origin/OpenSSL_1_1_1-stable
HEAD is now at a88ea7dfdf Check ASN1_item_ndef_i2d() return value.
romen@buildripper:/tmpram/111$ srcdir=$(realpath .); echo $srcdir
/tmpram/111
romen@buildripper:/tmpram/111$ blddir=../_build/nest/a/little/more
romen@buildripper:/tmpram/111$ mkdir -p $blddir
romen@buildripper:/tmpram/111$ cd $blddir
romen@buildripper:/tmpram/_build/nest/a/little/more$ $srcdir/config
Operating system: x86_64-whatever-linux2
Configuring OpenSSL version 1.1.1k-dev (0x101010b0L) for linux-x86_64
Using os-specific seed configuration
Creating configdata.pm
Creating Makefile

**********************************************************************
***                                                                ***
***   OpenSSL has been successfully configured                     ***
***                                                                ***
***   If you encounter a problem while building, please open an    ***
***   issue on GitHub <https://github.com/openssl/openssl/issues>  ***
***   and include the output from the following command:           ***
***                                                                ***
***       perl configdata.pm --dump                                ***
***                                                                ***
***   (If you are new to OpenSSL, you might want to consult the    ***
***   'Troubleshooting' section in the INSTALL file first)         ***
***                                                                ***
**********************************************************************
romen@buildripper:/tmpram/_build/nest/a/little/more$ make update
( cd ../../../../../111; /usr/bin/perl VMS/VMSify-conf.pl \
                        < apps/openssl.cnf > apps/openssl-vms.cnf )
( cd ../../../../../111; /usr/bin/perl crypto/bn/bn_prime.pl > crypto/bn/bn_prime.h )
( cd ../../../../../111; /usr/bin/perl crypto/objects/objects.pl -n \
                        crypto/objects/objects.txt \
                        crypto/objects/obj_mac.num \
                        > crypto/objects/obj_mac.new && \
    mv crypto/objects/obj_mac.new crypto/objects/obj_mac.num )
( cd ../../../../../111; /usr/bin/perl crypto/objects/objects.pl \
                        crypto/objects/objects.txt \
                        crypto/objects/obj_mac.num \
                        > include/openssl/obj_mac.h )
( cd ../../../../../111; /usr/bin/perl crypto/objects/obj_dat.pl \
                        include/openssl/obj_mac.h \
                        > crypto/objects/obj_dat.h )
( cd ../../../../../111; /usr/bin/perl crypto/objects/objxref.pl \
                        crypto/objects/obj_mac.num \
                        crypto/objects/obj_xref.txt \
                        > crypto/objects/obj_xref.h )
( cd ../../../../../111; /usr/bin/perl crypto/conf/keysets.pl \
                        > crypto/conf/conf_def.h )
( cd ../../../../../111; /usr/bin/perl crypto/asn1/charmap.pl \
                        > crypto/asn1/charmap.h )
( cd ../../../../../111; /usr/bin/perl fuzz/mkfuzzoids.pl \
                        crypto/objects/obj_dat.h \
                        > fuzz/oids.txt )
( b=`pwd`; set -e; cd ../../../../../111; \
          /usr/bin/perl util/ck_errf.pl -strict -internal; \
          /usr/bin/perl -I$b util/mkerr.pl  -internal )
( b=`pwd`; set -e; cd ../../../../../111/engines; \
          for E in *.ec ; do \
              /usr/bin/perl ../util/ck_errf.pl -strict \
                -conf $E `basename $E .ec`.c; \
              /usr/bin/perl -I$b ../util/mkerr.pl  -static \
                -conf $E `basename $E .ec`.c ; \
          done )
( b=`pwd`; cd ../../../../../111; /usr/bin/perl -I$b util/mkdef.pl crypto update )
Can't open opensslv.h at util/mkdef.pl line 1497.
make: *** [Makefile:597: ordinals] Error 2
```

**Comment by romen** [comment]
@levitte for convenience I created a PR that highlights the issue: #14388

**Comment by levitte** [comment]
Try this for a fix:

``` diff
diff --git a/Configurations/unix-Makefile.tmpl b/Configurations/unix-Makefile.tmpl
index 3a24d55135..41648c9526 100644
--- a/Configurations/unix-Makefile.tmpl
+++ b/Configurations/unix-Makefile.tmpl
@@ -917,8 +917,8 @@ errors:
           done )
 
 ordinals:
-	( b=`pwd`; cd $(SRCDIR); $(PERL) -I$$b util/mkdef.pl crypto update )
-	( b=`pwd`; cd $(SRCDIR); $(PERL) -I$$b util/mkdef.pl ssl update )
+	$(PERL) $(SRCDIR)/util/mkdef.pl crypto update
+	$(PERL) $(SRCDIR)/util/mkdef.pl ssl update
 
 test_ordinals:
        ( cd test; \
```

**Comment by romen** [comment]
> Try this for a fix:
> 
> ```diff
> diff --git a/Configurations/unix-Makefile.tmpl b/Configurations/unix-Makefile.tmpl
> index 3a24d55135..41648c9526 100644
> --- a/Configurations/unix-Makefile.tmpl
> +++ b/Configurations/unix-Makefile.tmpl
> @@ -917,8 +917,8 @@ errors:
>            done )
>  
>  ordinals:
> -	( b=`pwd`; cd $(SRCDIR); $(PERL) -I$$b util/mkdef.pl crypto update )
> -	( b=`pwd`; cd $(SRCDIR); $(PERL) -I$$b util/mkdef.pl ssl update )
> +	$(PERL) $(SRCDIR)/util/mkdef.pl crypto update
> +	$(PERL) $(SRCDIR)/util/mkdef.pl ssl update
>  
>  test_ordinals:
>         ( cd test; \
> ```

@levitte this patch does not apply, is this for 1.1.1?

**Comment by levitte** [comment]
> Ignore me, I was on the wrong branch!

I have no idea what you're talking about...  you applied it to the branch where you demonstrated the failure, and ta-daaaaa, it works!  I'm not sure I see what was wrong with that.

**Comment by romen** [comment]
> > Ignore me, I was on the wrong branch!
> 
> I have no idea what you're talking about... you applied it to the branch where you demonstrated the failure, and ta-daaaaa, it works! I'm not sure I see what was wrong with that.

I edited my previous comment: it said that you patch did not apply (but I was wrong!)

**Comment by romen** [comment]
Anyway, yes, your patch fixes the problem.

I am working to transform #14388 into the solution for this problem.

**Comment by romen** [comment]
No idea why this wasn't automagically closed by 01b9e18fb2!

**Comment by mattcaswell** [comment]
> No idea why this wasn't automagically closed by 01b9e18!

Auto closing only works on the master branch.
